### ISSUE TITLE
[no-unnecessary-type-assertion] Cast removed when it was necessary to infer a generic argument

### ISSUE THREAD

**Issue by jtbandes** [typescript-eslint/typescript-eslint#3502]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```ts
function foo<T>(): T {
  throw new Error();
}
function bar() {
  const x = foo() as string;
}
```

eslint config: https://github.com/foxglove/eslint-plugin/blob/main/configs/typescript.js
tsconfig: https://github.com/foxglove/studio/blob/main/packages/tsconfig/tsconfig.base.json

**Expected Result**

Without the `as string`, `x`'s type is `unknown`. With it, `T`, and therefore the type of `x`, is inferred to be `string`. So the cast actually has an effect, and should not be removed.

Having a function whose generic argument is only used as the return type and not any of the argument types is probably bad practice, but nonetheless it is valid TS code and should be handled correctly.

**Actual Result**

"This assertion is unnecessary since it does not change the type of the expression."

The no-unnecessary-type-assertion rule removes `as string`.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.26.1` |
| `@typescript-eslint/parser`        | `4.26.0` |
| `TypeScript`                       | `4.3.2` |
| `ESLint`                           | `7.27.0` |
| `node`                             | `15.9.0` |

**Comment by bradzacher** [comment]
There's unfortunately nothing we can do about this.
We cannot detect this case in any way, shape or form.

The tl;dr is that when a generic is ONLY used in an output position, TS will allow a type assertion to "backfill" its value.

So to you, the developer, it appears like `T` with no declaration would implicitly be typed as `unknown`, and then the assertion is converting an `unknown` to `string.
But in actual fact what's happening is that the `as string` causes TS to type `T` as `string`.
So when we inspect the types we see that the generic `T` is of type `string` and that the assertion is of type `string`. Thus it is deemed "unnecessary".

We've spoken with the TS team about this and it's "working as intended" from the TS side.

This is the folly of using a generic only in the return type position.
Using generics only in a return type position is an anti-pattern which should be avoided at all costs (yes - this is the response we got from the TS team).

**Comment by jtbandes** [comment]
Thanks for the response!

Wouldn’t it be possible to detect that the original function declaration had a generic parameter only in the output position? Or is that information somehow totally erased by TS?

It seems like false negatives (no warning where there could’ve been one) might be preferable to false positives here.

On the other hand, it’s not too hard to work around this issue by specifying the generic param within `<>` rather than a cast.
